src/tags/models.py

```diff
@@ -1,7 +1,38 @@
+from django.apps import apps
 from django.db import models
+from django.db.models import Sum
 
 from taggit.managers import TaggableManager
 
+from siteconfig.models import SiteConfig
+
+
+def total_xp_by_tags(user, tags: list):
+    """ 
+        Returns user's total xp over active_semester filtered by tags
+
+        expects a list of tags for tags arg (try list(tags))
+    """ 
+    # get models through here to prevent circular imports
+    QuestSubmission = apps.get_model("quest_manager", "QuestSubmission")
+    BadgeAssertion = apps.get_model("badges", "BadgeAssertion")
+
+    # get all objects with xp that is related to user and tag
+    xp_list = [
+        # QuestSubmission xp sum  
+        QuestSubmission.objects.all_approved(user, active_semester_only=True).filter(
+            is_completed=True, 
+            quest__tags__name__in=tags
+        ).exclude(do_not_grant_xp=True).distinct().aggregate(Sum("quest__xp"))['quest__xp__sum'],
+        
+        # BadgeAssertion xp sum 
+        BadgeAssertion.objects.all_for_user(user).filter(
+            semester=SiteConfig.get().active_semester, 
+            badge__tags__name__in=tags
+        ).exclude(do_not_grant_xp=True).distinct().aggregate(Sum("badge__xp"))['badge__xp__sum'],
+    ]
+    return sum(xp for xp in xp_list if xp is not None)
```

Comment:
With your new version, would the sum ever be None? If not, I think it would be simpler to get rid of the list and do

`return quest_xp + badge_xp`

Reply:
sum can be none since ``aggregate()`` will always return a dictionary of keys of whatever arg you put in.
if there are no badges or quests linked to the user, aggregate will return a dictionary like: ``{ 'badge__xp__sum': None }``

Reply:
i could write 
``
return( quest_xp or 0) + (badge_xp or 0)
``
but it looks ugly IMO

Reply:
Can you paste this in as a code comment?

```
sum can be none since aggregate() will always return a dictionary of keys of whatever arg you put in.
if there are no badges or quests linked to the user, aggregate will return a dictionary like: { 'badge__xp__sum': None }
```

Reply:
done